core/src/main/java/io/questdb/griffin/engine/groupby/GroupByUtils.java

```diff
@@ -100,13 +107,79 @@ public static void assembleGroupByFunctions(
     ) throws SqlException {
         try {
             outGroupByFunctionPositions.clear();
-            outRecordFunctionPositions.clear();
+            projectionFunctionPositions.clear();
             int fillCount = sampleByFill != null ? sampleByFill.size() : 0;
 
             int columnKeyCount = 0;
             int lastIndex = -1;
             final ObjList<QueryColumn> columns = model.getColumns();
 
+            // compile functions upfront and assemble the metadata for group-by
+            for (int i = 0, n = columns.size(); i < n; i++) {
+                final QueryColumn column = columns.getQuick(i);
+                final ExpressionNode node = column.getAst();
+                final int index = baseMetadata.getColumnIndexQuiet(node.token);
+                TableColumnMetadata m = null;
+                if (node.type != LITERAL || index != timestampIndex || timestampUnimportant) {
+                    Function func = functionParser.parseFunction(
+                            node,
+                            baseMetadata,
+                            executionContext
+                    );
+
+                    // functions added to the outer projections will later be replaced by column refences
+                    outerProjectionFunctions.add(func);
```

Comment:
In the outer `catch` block we have this:
```java
            Misc.freeObjListAndClear(outerProjectionFunctions);
            Misc.freeObjList(outGroupByFunctions);
            Misc.freeObjList(innerProjectionFunctions);
```

Later on, we may add functions from `outerProjectionFunctions` to `outGroupByFunctions`, so the same function may be present in all three lists.

So, we're adding the same function to different lists twice leading to potential double (or to be more precise, triple) close. Is that on purpose?

Reply:
indeed

Reply:
cleaned up in [26779f4](https://github.com/questdb/questdb/pull/5778/commits/26779f4614c3a920102fe695d2c491901c70ad7e) outerFunctions contain all others that are not columns.